lib/model/model.go

```diff
@@ -2306,26 +2407,98 @@ func (m *model) AddConnection(conn protocol.Connection, hello protocol.Hello) {
 	}
 
 	m.deviceWasSeen(deviceID)
+	m.scheduleConnectionPromotion()
+}
+
+func (m *model) scheduleConnectionPromotion() {
+	m.promotionTimer.Reset(time.Second)
+}
+
+// promoteConnections checks for devices that have connections, but where
+// the primary connection hasn't started index handlers etc. yet, and
+// promotes the primary connection to be the index handling one. This should
+// be called after adding new connections, and after closing a primary
+// device connection.
+func (m *model) promoteConnections() {
+	m.fmut.RLock() // for generateClusterConfigFRLocked called by promoteDeviceConnectionLocked
+	defer m.fmut.RUnlock()
+
+	m.pmut.Lock() // for most other things
+	defer m.pmut.Unlock()
+
+	for deviceID, connIDs := range m.deviceConnIDs {
+		// Figure out the best current connection priority for this device.
+		bestPriority := m.connections[connIDs[0]].Priority()
+		for _, connID := range connIDs[1:] {
+			priority := m.connections[connID].Priority()
+			if priority < bestPriority {
+				bestPriority = priority
+			}
+		}
+
+		// Close connections with a worse connection priority than best.
```

Comment:
This feels like something the connection service can and should handle when adding new connections. Couldn't we expose all connections on a `Model.Connections` method which then also removes the need for the new `deviceConnectionCounter` in `connections.service`?

Reply:
I think having the model track connections (for the purpose of external consumers) was a misdesign and I was happy to remove it. (Tell, don't ask, etc.) That said, possibly the connection service could do this. I will look into it.

Reply:
Still thinking about this one. Certainly, the connection service could track all connections and "reap" ones with worse priority after a while, thus triggering a promotion of some other connection when the model reacts to the close...